account_cash_discount_payment/models/account_payment_line.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+# Copyright 2014 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, models, fields
+
+
+class PaymentLine(models.Model):
+
+    _inherit = 'account.payment.line'
+
+    pay_with_discount = fields.Boolean(
+        default=False,
+    )
+    discount_due_date = fields.Date(
+        compute='_compute_discount_due_date',
+    )
+    discount_amount = fields.Monetary(
+        default=0.0
+    )
+    original_amount_currency = fields.Monetary(
+        readonly=True,
+    )
```

Comment:
Can't we do the economy of this field? I think we can rely on the invoice total and enable the cash discount payment only if the invoice residual is equal to the invoice total. Ie if there is no invoice, disable pay with discount, and if the discount is partially paid, disable pay with discount. @chrisdec